app/services/puppet_fact_parser.rb

```diff
@@ -145,6 +145,29 @@ def boot_timestamp
     uptime_seconds.nil? ? nil : (Time.zone.now.to_i - uptime_seconds.to_i)
   end
 
+  def virtual
+    facts['is_virtual']
+  end
+
+  def ram
+    facts['memorysize']
+    if facts.key?('memory')
+      value = facts.fetch('memory', {}).fetch('system', {}).fetch('total_bytes', nil)
+      return if value.nil?
+      value / 1.megabyte
+    else
+      facts['memorysize_mb']
+    end
+  end
+
+  def sockets
+    facts['physicalprocessorcount'] || facts.fetch('processors', {}).fetch('physicalcount', nil)
```

Comment:
Let's prefer modern facts over legacy ones and use dig

```suggestion
    facts.dig('processors', 'physicalcount') || facts['physicalprocessorcount']
```